src/daft-io/src/azure_blob.rs

```diff
@@ -117,6 +139,253 @@ impl AzureBlobSource {
+        let prefix_with_delimiter = format!("{}{delimiter}", prefix.trim_end_matches(&delimiter));
+        let full_path = format!("{}://{}{}", protocol, container_name, prefix);
+        let full_path_with_trailing_delimiter = format!(
+            "{}://{}{}",
+            protocol, container_name, &prefix_with_delimiter
+        );
+
+        let mut unchecked_results = self
+            ._list_directory_delimiter_stream(
+                &container_client,
+                &protocol,
+                &container_name,
+                &prefix_with_delimiter,
+                &delimiter,
+            )
+            .await;
+
+        // The Azure API call result is almost identical to the desired list directory result
+        // with a couple of exceptions:
+        // 1. Nonexistent paths return an empty list instead of 404.
+        // 2. A file "path" can be returned for a lookup "path/", which is not the desired behaviour.
+        //
+        // To check for and deal with these cases,
+        // manually process the first two items of the stream.
+
+        let mut maybe_first_two_items = vec![];
+        let mut stream_exhausted = false;
+        for _ in 0..2 {
+            let item = unchecked_results.next().await;
+            if let Some(item) = item {
+                maybe_first_two_items.push(item);
+            } else {
+                stream_exhausted = true;
+                break;
+            }
+        }
+
+        // Make sure the stream is pollable even if empty,
+        // since we will chain it later with the two items we already popped.
+        let unchecked_results = if !stream_exhausted {
+            unchecked_results
+        } else {
+            futures::stream::iter(vec![]).boxed()
+        };
+
+        match &maybe_first_two_items[..] {
+            [] => {
+                // Check whether the path actually exists.
+                let s = async_stream::stream! {
+                    let prefix_exists = match prefix.as_str() {
+                        "" | "/" => true,
+                        _ => {
+                            // To check whether the prefix actually exists, check whether it exists as a result one directory above.
+                            // (Azure does not return marker files for empty directories.)
+                            let upper_dir = prefix // "/upper/blah/"
+                                .trim_end_matches(&delimiter) // "/upper/blah"
+                                .trim_end_matches(|c: char| c.to_string() != delimiter); // "/upper/"
+
+                            let upper_results = container_client
+                                .list_blobs()
+                                .delimiter(delimiter.to_string())
+                                .prefix(upper_dir.to_string())
+                                .into_stream()
+                                .try_collect::<Vec<_>>()
```

Comment:
You shouldnt have to collect this and then iterate over it. You should be able to use [`then`](https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.then) to transform (map) and then [`any`](https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.any) to terminate as soon as it finds the condition of `(filepath == full_path_with_trailing_delimiter)`.